### ISSUE TITLE
NativeAOT does not honor new() constraint

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtime#80037]
### Description

NativeAOT does not honor `new()` constraint applied to interface. 

### Reproduction Steps

```csharp
using System.Diagnostics.CodeAnalysis;

var serviceCollection = new ServiceCollection();
var w = serviceCollection.GetRequiredService<IGeneric<Target>>();
Console.WriteLine("Working");


class ServiceCollection
{
    public T GetRequiredService<T>()
    {
        return (T)Activator.CreateInstance(typeof(GenericInstantiaion<>).MakeGenericType(typeof(T).GenericTypeArguments[0]))!;
    }
}

interface IGeneric<T> where T: new() {}
class GenericInstantiaion<T> : IGeneric<T> where T: new()
{
    public GenericInstantiaion()
    {
        new T();
    }
}

public class Target {

}

```

### Expected behavior

Do not trim constructor of Target class.

### Actual behavior

Exception at runtime.

### Regression?

No

**Comment by MichalStrehovsky** [comment]
Thanks! Here's a repro that doesn't generate trimming warnings (and is also broken):

```csharp
IGeneric<Target>.Hey();


interface IGeneric<T> where T: new()
{
    public static void Hey()
    {
        Activator.CreateInstance(typeof(T));
    }

}

public class Target { }
```

**Comment by kant2002** [comment]
I think my sample closer to how open generic registered in MEDI. Your example probably different issue :)

**Comment by MichalStrehovsky** [comment]
The issue would be a won't fix if it required MakeGeneric, especially if it's with something that has a new constraint. If there's a warning, trimming can and will generate broken things. The only aspects that are analysis holes are if there's no warnings.

**Comment by kant2002** [comment]
I do not want replicate whole DI setup. That’s bigger sample with only one warning from BuildServiceProvider which I believe is fine.
```csharp
using System.Diagnostics.CodeAnalysis;
using Microsoft.Extensions.DependencyInjection;

var serviceCollection = new ServiceCollection();
serviceCollection.AddSingleton(typeof(IGeneric<>),typeof(GenericInstantiaion<>));
var serviceProvider = serviceCollection.BuildServiceProvider();
var w = serviceProvider.GetRequiredService<IGeneric<Target>>();
Console.WriteLine("Working");




interface IGeneric<T> where T: new() {}
class GenericInstantiaion<T> : IGeneric<T> where T: new()
{
    public GenericInstantiaion()
    {
        new T();
    }
}

public class Target {

}
```

**Comment by kant2002** [comment]
We can have situations like in MEDI where library author would support similar pattern for construction as in medi for open generics and just suppress warning like here https://github.com/dotnet/runtime/blob/199436b51fba6f0bee469c63ee42fa266d70222d/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L349-L353

I think it's reasonable to perform analysys on parameters and return types of method when they specified. Develop specify that he need type, and that type guaranteed to return specific type by compiler, so that means that constraints on specific generic type would be satisfied. I honestly did not see how developer can break something with this extra requirement.

Yes, internally it's `MakeGenericType` probably, but why I cannot suppress usage if I know what I'm doing?

**Comment by MichalStrehovsky** [comment]
> I do not want replicate whole DI setup. That’s bigger sample with only one warning from BuildServiceProvider which I believe is fine.

Oh, I see, it's the pattern from DI. Yes, the suppression in DI relies on the fact that because the interface has a `new()` constraint and there's a concrete instantiation of the interface with type `X`, we will not trim the constructor of `X`. The reduced sample will cover that and it is the same issue. The compiler is not great at detecting all the places where generic instantiations happen.

**Comment by MichalStrehovsky** [comment]
Vitek fixed this in #82818.
